media_driver/linux/common/os/i915/mos_bufmgr.c

```diff
@@ -4691,3 +4691,246 @@ mos_bufmgr_gem_init(int fd, int batch_size)
+}
+
+int mos_query_engines(int fd,
+                      __u16 engine_class,
+                      __u64 caps,
+                      unsigned int *nengine,
+                      struct i915_engine_class_instance *ci)
+{
+    struct drm_i915_query query;
+    struct drm_i915_query_item query_item;
+    struct drm_i915_query_engine_info *engines;
+    int ret, len;
+
+    engines = (drm_i915_query_engine_info *)malloc(4096);
+    memset(engines,0,4096);
+    memclear(query_item);
+    query_item.query_id = DRM_I915_QUERY_ENGINE_INFO;
+    query_item.length = 0;
+    query_item.data_ptr = (uintptr_t)engines;
+    memclear(query);
+    query.num_items = 1;
+    query.items_ptr = (uintptr_t)&query_item;
+
+    ret = drmIoctl(fd, DRM_IOCTL_I915_QUERY, &query);
+    if (ret)
+    {
+        goto fini;
+    }
+    len = query_item.length;
+
+    memset(engines,0,4096);
+    memclear(query_item);
+    query_item.query_id = DRM_I915_QUERY_ENGINE_INFO;
+    query_item.length = len;
+    query_item.data_ptr = (uintptr_t)engines;
+    memclear(query);
+    query.num_items = 1;
+    query.items_ptr = (uintptr_t)&query_item;
+
+    ret = drmIoctl(fd, DRM_IOCTL_I915_QUERY, &query);
+    if (ret)
+    {
+        goto fini;
+    }
+
+    int i, num;
+    for (i = 0, num = 0; i < engines->num_engines; i++) {
+        struct drm_i915_engine_info *engine =
+            (struct drm_i915_engine_info *)&engines->engines[i];
+        if ( engine_class == engine->engine_class
+             && ((caps & engine->capabilities) || caps == 0))
+        {
+            ci->engine_class = engine_class;
+            ci->engine_instance = engine->engine_instance;
+            ci++;
+            num++;
+        }
+    }
+    if (num > *nengine)
+    {
+        fprintf(stderr,"%s: Number of engine instances out of range, %d,%d\n",
+                __FUNCTION__, num, *nengine);
+        goto fini;
+    }
+    *nengine = num;
+
+
+fini:
+    free(engines);
+    return ret;
+}
+
+int mos_set_context_param_load_balance(struct mos_linux_context *ctx,
+                     const struct i915_engine_class_instance *ci,
+                     unsigned int count)
+{
+    int ret;
+    uint32_t size;
+
+    assert(ci);
+
+    /*
+    struct i915_context_engines_load_balance {
+	struct i915_user_extension base;
+
+	__u16 engine_index;
+	__u16 num_siblings;
+	__u32 flags;
+
+	__u64 mbz64;
+
+	struct i915_engine_class_instance engines[count];
+    }balancer;
+    */
+    size = sizeof(struct i915_context_engines_load_balance) + count * sizeof(*ci);
+    uint32_t* balancer = (uint32_t*)malloc(size);
+    if (NULL == balancer)
+    {
+        return -ENOMEM;
+    }
+    memset(balancer, 0, size);
+    ((struct i915_context_engines_load_balance*) balancer)->base.name = I915_CONTEXT_ENGINES_EXT_LOAD_BALANCE;
+    ((struct i915_context_engines_load_balance*) balancer)->num_siblings = count;
+    memcpy(((struct i915_context_engines_load_balance*) balancer)->engines, ci, count * sizeof(*ci));
+
```

Comment:
Why is balancer a pointer to uint32_t? If it was a pointer to struct i915_context_engines_load_balance then you wouldn't need all these casts.

Reply:
fixed.